### DISCUSSION TITLE
Loading all List Items

### DISCUSSION THREAD

**Discussion by bajce** [pnp/pnpcore#579] [Q&A] [answered]
I want to get all Items and its RoleAssignments with RoleDefinitions with less possible server round trips and have following asynchron code for that: 

```
var list = context.Web.Lists.GetById(new Guid('xxxx'), l => l.Items.QueryProperties(
                    item => item.UniqueId,
                    item => item.RoleAssignments.QueryProperties(
                        ra => ra.PrincipalId, 
                        ra => ra.RoleDefinitions.QueryProperties(rd => rd.Id, rd => rd.Name))
                ));
```
If the List has more the 100 Items I have to use paging:
In order to load all Items I use `list.Items.ToList()`. However this even do I defined with Parameter expression which properties should be loaded, they are not.  I guess I have to define that before ToList() again and I guess I have to repeat that for each nested collection property?

What would be the best approach to load all Items with their RoleAssignments and RoleDefinitions with minimal server round trips?

**Comment by jansenbe** [comment, marked as answer]
@bajce : you can't do that in one go as you can have bump in to very large lists + there's the extra complexity of not being able to filter on ListItem.HasUniqueRoleAssignments. You'll have to request all list items and evaluate client side. Once you've built up a list of items with unique permissions you can then build a batch and send back that batch to get the actual defined role assignments. Below sample shows this. 

Note that that below approach automatically pages the data loads per 100 items. You can change the 100 by changing the default via configuration (see https://pnp.github.io/pnpcore/using-the-sdk/basics-settings.html) or by using `Take()` and `Skip()` in your code (see https://pnp.github.io/pnpcore/using-the-sdk/basics-getdata-paging.html#paging-via-the-takeskip-extension-methods).

```csharp
await foreach (var list in context2.Web.Lists/*.Where(item => item.HasUniqueRoleAssignments == true)*/
                                             .QueryProperties(list => list.HasUniqueRoleAssignments, 
                                                              list => list.Title).AsAsyncEnumerable())
{
    
    List<IListItem> listItems = new List<IListItem>();

    // Filter on HasUniqueRoleAssignments does not work for ListItem, not using REST, not using CAML query
    await foreach (var item in list.Items/*.Where(item => item.HasUniqueRoleAssignments == true)*/
                                         .QueryProperties(p => p.HasUniqueRoleAssignments)
                                         .AsAsyncEnumerable())
    {
        if (item.HasUniqueRoleAssignments)
        {
            context2.Logger.LogInformation($"{list.Title} - {item.Id}");
            listItems.Add(item);
        }
    }

    if (listItems.Count > 0)
    {
        var itemDetailsBatch = context2.NewBatch();
        foreach (var item in listItems)
        {
            await item.LoadBatchAsync(itemDetailsBatch, 
                                      item => item.RoleAssignments.QueryProperties(
                                                    ra => ra.PrincipalId,
                                                    ra => ra.RoleDefinitions.QueryProperties(rd => rd.Id, rd => rd.Name)));
        }
        await context2.ExecuteAsync(itemDetailsBatch);

        foreach(var item in listItems)
        {
            // Process the list items with their custom role assignments...
        }
    }

```

**Reply by jansenbe** [reply]
@bajce : I've added the missing documentation for working with custom permissions (https://pnp.github.io/pnpcore/using-the-sdk/security-intro.html) + I've unified the implementation for Web, List and ListItem: all now implement an `ISecurableObject` interface which brings along the needed properties and methods.

**Reply by bajce** [reply]
Thanks for the answers. I just tried the approach from above and it looks like all the Items are loaded, but I have couple questions about this approach:

1. Will all RoleAssignments and RoleDefinitions be loaded using this approach (if I have more RoleAssignments or RoleDefinitions then max paginator size)? 

2. Applying same approach for Web.Webs.QueryProperties(...) does not load the properties. Is there something different? Will paging be applied for SubWebs at all?

3. How to load additional ExpandDataProperties from Item e.g item["FileLeafRef"] (because GetDisplayName makes extra request) item["ParentUniqueId]) (because in want to get build the folder structure so I need to somehow map the parent). I tried to put it as selector in QueryProperties but it did't loaded it.

4. Is it possible to load the "direct" items e.g if I have huge tree structures with folders, and I want to load only direct sub folders, process that and load the items of sub folders and so on.

**Reply by jansenbe** [reply]
Hi @bajce :

About your questions:

1. Yes, they should all be loaded, there's no paging used for properties loaded using Get/Load methods
2. You need a PnPContext per web, you can't get data from different webs in a single API call. See https://pnp.github.io/pnpcore/using-the-sdk/webs-intro.html#getting-a-web for how to iterate over the webs in a site
3. Correct, in that case it's better to replace `await foreach (var item in list.Items...` with a paged `ListDataAsStream` approach as shown here https://pnp.github.io/pnpcore/using-the-sdk/listitems-intro.html#using-paging-with-listdataasstream
4. Yes, using the `ListDataAsStream` approach you set the server relative url of the folder you want to query via the `FolderServerRelativeUrl` property of the `RenderListDataOptions` class

**Reply by bajce** [reply]
@jansenbe  Thank You. 

3. I just tried that, but for some reason properties like HasUniqueRoleAssignments are not loaded. I also tried to add that in ViewFields of Qeury (<FieldRef Name='HasUniqueRoleAssignments' />) but didn't change anything. As a workaround im loading 'SharedWithDetails' FieldRef and using following statement to evaluate HasUniqueRoleAssignments: item.Values.ContainsKey("SharedWithDetails"). Is this approach ok or is there any better and more elegant solution?

**Reply by jansenbe** [reply]
@bajce : Yesterday I've done some changes that allows to use LoadItemsByCamlQuery in combination with loading "system fields" (like `FileLeafRef`) and expand other list item collection (see https://pnp.github.io/pnpcore/using-the-sdk/listitems-intro.html) and added instructions on how to choose the best list item retrieval option. I think you'll find the sample in https://pnp.github.io/pnpcore/using-the-sdk/listitems-intro.html#using-paging-with-loaditemsbycamlquery very fitting for your purpose :-)

**Reply by bajce** [reply]
@jansenbe That's exactly what I need. Thank you very much.
